src/System.IO.Compression/src/System/IO/Compression/DeflateStream.cs

```diff
@@ -18,124 +18,106 @@ public partial class DeflateStream : Stream
 
                 case CompressionMode.Compress:
-                    if (!stream.CanWrite)
-                    {
-                        throw new ArgumentException(SR.NotWriteableStream, "stream");
-                    }
-                    _deflater = CreateDeflater(null);
+                    InitializeDeflater(stream, leaveOpen, windowBits, CompressionLevel.Optimal);
                     break;
 
                 default:
                     throw new ArgumentException(SR.ArgumentOutOfRange_Enum, "mode");
             }
-
-            _stream = stream;
-            _mode = mode;
-            _leaveOpen = leaveOpen;
-            _buffer = new byte[DefaultBufferSize];
-        }
-
-        // Implies mode = Compress
-        public DeflateStream(Stream stream, CompressionLevel compressionLevel)
-
-            : this(stream, compressionLevel, false)
-        {
         }
 
-        // Implies mode = Compress
-        public DeflateStream(Stream stream, CompressionLevel compressionLevel, bool leaveOpen)
+        /// <summary>
+        /// Internal constructor to specify the compressionlevel as well as the windowbits
+        /// </summary>
+        internal DeflateStream(Stream stream, CompressionLevel compressionLevel, bool leaveOpen, int windowBits)
         {
             if (stream == null)
                 throw new ArgumentNullException("stream");
 
-            if (!stream.CanWrite)
-                throw new ArgumentException(SR.NotWriteableStream, "stream");
+            InitializeDeflater(stream, leaveOpen, windowBits, compressionLevel);
+        }
+
+        /// <summary>
+        /// Sets up this DeflateStream to be used for Zlib Inflation/Decompression
+        /// </summary>
+        internal void InitializeInflater(Stream stream, bool leaveOpen, int windowBits)
+        {
+            Debug.Assert(windowBits >= -15 && windowBits <= 47);
+            Debug.Assert(stream != null);
+            if (!stream.CanRead)
+                throw new ArgumentException(SR.NotReadableStream, "stream");
 
-            Contract.EndContractBlock();
+            _inflater = new Inflater(windowBits);
 
             _stream = stream;
-            _mode = CompressionMode.Compress;
+            _mode = CompressionMode.Decompress;
             _leaveOpen = leaveOpen;
-
-            _deflater = CreateDeflater(compressionLevel);
-
             _buffer = new byte[DefaultBufferSize];
         }
 
-        private static Deflater CreateDeflater(CompressionLevel? compressionLevel)
+        /// <summary>
+        /// Sets up this DeflateStream to be used for Zlib Deflation/Compression
+        /// </summary>
+        internal void InitializeDeflater(Stream stream, bool leaveOpen, int windowBits, CompressionLevel compressionLevel)
         {
-            return compressionLevel.HasValue ?
-                new Deflater(compressionLevel.Value) :
-                new Deflater();
-        }
+            Debug.Assert(windowBits >= -15 && windowBits <= 31);
```

Comment:
Nit: consider putting these values into meaningfully-named constants (they can be local consts).